### ISSUE TITLE
userData is sometimes not available when page is loaded (during testing)

### ISSUE THREAD

**Issue by djuber** [forem/forem#13228]
**Describe the bug**

https://github.com/forem/forem/pull/13224 adapted a failing test that expected a component to have rendered with one element disabled. It was observed that loading the page twice "solved" the issue (or, the tests began to pass reliably once this was done).

The underlying issue appears to be that the user has not been added by the front end when the system test views the page (the backend/rails does have a logged in user, but the code to remove elements not accessible to non-admin or non-trusted users did not run in time.

**To Reproduce**

Check out https://github.com/forem/forem/commit/28e38cbb7d03b635c067f1b41a82cbfb46f844fd (or remove the `2.times` in the current "Flagging users from profile page" spec spec/system/user/trusted_user_flags_user_spec.rb:16

```
HEADLESS=false bundle exec rspec --format=documentation --order=random --seed=9374
```

Test for when logged in as a non-trusted user fails.

If you add a breakpoint(`binding.pry`) after `click_button(id: "user-profile-dropdown")` and before any expectations, the browser will show the erroneous situation (you're logged in as a non-privileged user, viewing the profile of another user, but the "Flag user" option is visible in the expanded dropdown).

**Expected behavior**

Flag user option was hidden. https://github.com/forem/forem/blob/master/app/javascript/profileDropdown/flagButton.js#L26-L29
```javascript
 if (!trustedOrAdmin || user.id === parseInt(profileUserId, 10)) {
    flagButton.remove();
  }
```

Testing suggested that the earlier guard clause was executed instead (so the removal of the flag 

```javascript
  const user = userData();
  if (!user) {
    // add a log statement to confirm this happens:
    console.log("user was missing."); 
    return;
  }
```

**Screenshots**

What the test saw but did not expect:

![showing-for-non-trusted-user](https://user-images.githubusercontent.com/1237369/113352869-7a89c900-9302-11eb-91ff-2242e22424b9.png)

The conditions that would have led to the removal were all set correctly by the time I had a console on the page - suggesting this could be a sequencing problem (when do pieces of JS get executed, and is that consistent/reliable):

![should-have-called-remove](https://user-images.githubusercontent.com/1237369/113352885-7d84b980-9302-11eb-9531-cefde9649a8e.png)

**Additional context**

@maestromac mentioned this situation (userData did not load correctly prior to viewing the page, causing a flaky test) had been a common problem in the past.

> This actually is the core problem to some of our already disabled flaky spec. Backend is logged in but frontend isn't. I never figured out why sometime userData() is nil but user is already signed-in on the top bar

**Comment by nickytonline** [comment]
Using a promised based utility function we have to get the user would probably make this code more reliable.

https://github.com/forem/forem/blob/009c14653ab84da91cf278400350503e056bef29/app/javascript/chat/util.js#L40-L44

It's currently in connect utilities, but we could add it to our utilities folder and reuse it here as well as connect. There are a few other places that use the connect util as well, so we'd need to refactor them as well.

![image](https://user-images.githubusercontent.com/833231/113454200-ae81ee80-93d5-11eb-81c9-96242a5c42c9.png)

**Comment by djuber** [comment]
@nickytonline Thanks for the tip - I spent a fair amount of yesterday getting to understand what was happening. I see the issue now is that we use local storage object (accessed via `browserStoreCache`) to cache the user info (and an async call to update base data during `initializeBodyData` will refresh this for logged in users on page load). The flaky test is around code that uses the DOM's copy of the user info (via the `userData` wrapper) to make decisions, which is populated from the local storage object early if available, and updated after the xhr is done and the callback resyncs state (so we are rendering the dropdown without a valid user in the frontend, while the test expected a logged in user). 

In practice this is fine (users are usually either logged in or logged out, local storage will be probably available), it's in the test case where we're regularly logging in, logging out, changing users, and quickly expecting content to be right on the first page view. The repeated execution in and out of a logged in state causes the LSO to get cleared (initializeBodyData will clear the current user from local storage if there's no user in base data), in ways that are unexpected in the wild (and caused the flakiness).

I wonder if there's not a way to provide minimal userData information when we splice in the top bar (the backend is injecting the logged in users information to the navbar dropdown) and read from that rather than `document.body.dataset.user` - I haven't looked under the hood at the /async_info/base_data to see whether loading username, user id, and role information early will be problematic (but it would solve the "sometimes userData is null during selenium driven tests" situation we started with). 

I'm still digesting all of this but it looks like the getWaitOnUserDataHandler logic is connected to app/assets/javascripts/initializers/initializeBodyData.js (or is waiting on it) - if what you're suggesting is we add wait for user data logic to the dropdown render this makes sense (and in that case promoting the chat utility code to a general utility makes sense).
